state/executor.go

```diff
@@ -0,0 +1,229 @@
+	// TODO(tzdybal): blockID
+	s := State{
+		Version:         state.Version,
+		ChainID:         state.ChainID,
+		InitialHeight:   state.InitialHeight,
+		LastBlockHeight: int64(block.Header.Height),
+		LastBlockTime:   time.Unix(int64(block.Header.Time), 0),
+		// skipped all "Validators" fields
+		ConsensusParams:                  state.ConsensusParams,
+		LastHeightConsensusParamsChanged: state.LastHeightConsensusParamsChanged,
+	}
+	copy(s.LastResultsHash[:], lltypes.NewResults(abciResponses.DeliverTxs).Hash())
+
+	return s, nil
+}
+
+func (e *BlockExecutor) commit(state State, block *types.Block, deliverTxs []*abci.ResponseDeliverTx) ([]byte, uint64, error) {
+	e.mempool.Lock()
+	defer e.mempool.Unlock()
+
+	err := e.mempool.FlushAppConn()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	resp, err := e.proxyApp.CommitSync(context.TODO())
+	if err != nil {
+		return nil, 0, err
+	}
+
+	maxBytes := state.ConsensusParams.Block.MaxBytes
+	maxGas := state.ConsensusParams.Block.MaxGas
+	err = e.mempool.Update(int64(block.Header.Height), fromOptimintTxs(block.Data.Txs), deliverTxs, mempool.PreCheckMaxBytes(maxBytes), mempool.PostCheckMaxGas(maxGas))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return resp.Data, uint64(resp.RetainHeight), err
+}
+
+func (e *BlockExecutor) validate(state State, block *types.Block) error {
+	if block.Header.Version.App != state.Version.Consensus.App ||
+		block.Header.Version.Block != state.Version.Consensus.Block {
+		return errors.New("block version mismatch")
+	}
+	if state.LastBlockHeight <= 0 && block.Header.Height != uint64(state.InitialHeight) {
+		return errors.New("initial block height mismatch")
+	}
+	if state.LastBlockHeight > 0 && block.Header.Height != uint64(state.LastBlockHeight)+1 {
+		return errors.New("block height mismatch")
+	}
+	if !bytes.Equal(block.Header.AppHash[:], state.AppHash[:]) {
+		return errors.New("AppHash mismatch")
+	}
+
+	if !bytes.Equal(block.Header.LastResultsHash[:], state.LastResultsHash[:]) {
+		return errors.New("LastResultsHash mismatch")
+	}
+
+	// TODO(tzdybal): what else to check
+
+	return nil
+}
+
+func (e *BlockExecutor) execute(state State, block *types.Block) (*tmstate.ABCIResponses, error) {
+	abciResponses := new(tmstate.ABCIResponses)
+	abciResponses.DeliverTxs = make([]*abci.ResponseDeliverTx, len(block.Data.Txs))
+
+	txIdx := 0
+	validTxs := 0
+	invalidTxs := 0
+
+	var err error
+
+	e.proxyApp.SetResponseCallback(func(req *abci.Request, res *abci.Response) {
+		if r, ok := res.Value.(*abci.Response_DeliverTx); ok {
+			txRes := r.DeliverTx
+			if txRes.Code == abci.CodeTypeOK {
+				validTxs++
+			} else {
+				e.logger.Debug("Invalid tx", "code", txRes.Code, "log", txRes.Log)
+				invalidTxs++
+			}
+			abciResponses.DeliverTxs[txIdx] = txRes
+			txIdx++
+		}
+	})
+
+	hash, err := hash.Hash(block)
+	if err != nil {
+		return nil, err
+	}
+	abciResponses.BeginBlock, err = e.proxyApp.BeginBlockSync(
+		context.TODO(),
```

Comment:
I think it would be a bonus if we could start handling all of these todo contexts. Would it make sense to use a context specific to the `*BlockExecutor` or to pass it as an argument?